integration/suites/harness/agent_test.go

```diff
@@ -0,0 +1,152 @@
+		brokerLogger  *logrus.Logger
+		brokerLogbuff *gbytes.Buffer
+		serverLogbuff *gbytes.Buffer
+		mockCtl       *gomock.Controller
+		err           error
+	)
+
+	startServerInstance := func(cfgFile string, logbuff *gbytes.Buffer) (*server.Instance, error) {
+		logger := logrus.New()
+		logger.SetOutput(logbuff)
+
+		srv, err = testutil.StartServerInstance(ctx, &wg, cfgFile, logger)
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(logbuff).Should(gbytes.Say("Connected to nats://localhost:4222"))
+
+		return srv, nil
+	}
+
+	createAgent := func(fw inter.Framework) *mcorpc.Agent {
+		ddl, err := fs.FS.ReadFile("ddl/cache/agent/rpcutil.json")
+		Expect(err).ToNot(HaveOccurred())
+
+		rpcutilAgent, err = agentharness.NewWithDDLBytes(fw, mockCtl, "rpcutil", ddl)
+		Expect(err).ToNot(HaveOccurred())
+
+		a, err := rpcutilAgent.Agent()
+		Expect(err).ToNot(HaveOccurred())
+
+		return a
+	}
+
+	createRpcUtilClient := func() (*gbytes.Buffer, *rpcutilclient.RpcutilClient, error) {
+		logBuff, logger := testutil.GbytesLogger(logrus.DebugLevel)
+
+		cfg, err := config.NewConfig("testdata/client.conf")
+		if err != nil {
+			return nil, nil, err
+		}
+
+		cfg.CustomLogger = logger
+		cfg.OverrideCertname = "rip.mcollective"
+
+		fw, err := choria.NewWithConfig(cfg)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		client, err := rpcutilclient.New(fw)
+		if err != nil {
+			return nil, nil, err
+		}
+
+		return logBuff, client, nil
+	}
+
+	BeforeEach(func() {
+		mockCtl = gomock.NewController(GinkgoT())
+
+		ctx, cancel = context.WithTimeout(context.Background(), 10*time.Second)
+		DeferCleanup(func() {
+			cancel()
+			Eventually(brokerLogbuff, 5).Should(gbytes.Say("Choria Network Broker shut down"))
+		})
+
+		brokerLogbuff, brokerLogger = testutil.GbytesLogger(logrus.DebugLevel)
+		serverLogbuff = gbytes.NewBuffer()
+
+		_, err := testbroker.StartNetworkBrokerWithConfigFile(ctx, &wg, "testdata/broker.conf", brokerLogger)
+		Expect(err).ToNot(HaveOccurred())
+		Eventually(brokerLogbuff, 1).Should(gbytes.Say("Server is ready"))
+
+		srv, err = startServerInstance("testdata/server.conf", serverLogbuff)
+		Expect(err).ToNot(HaveOccurred())
+
+		err = srv.AgentManager().RegisterAgent(ctx, "rpcutil", createAgent(srv.Choria()), srv.Connector())
+		Expect(err).ToNot(HaveOccurred())
+
+		Eventually(serverLogbuff).Should(gbytes.Say("Registering new agent rpcutil of type rpcutil"))
+
+		_, rpcutilClient, err = createRpcUtilClient()
+		Expect(err).ToNot(HaveOccurred())
+	})
+
+	AfterEach(func() {
+		mockCtl.Finish()
+	})
+
+	Describe("action stubbing", func() {
+		It("Should do the correct stubs", func() {
+			rpcutilAgent.Stub("ping", func(_ context.Context, _ *mcorpc.Request, reply *mcorpc.Reply, _ *mcorpc.Agent, _ inter.ConnectorInfo) {
+				reply.Data = map[string]interface{}{
+					"pong": time.Now().Unix(),
+				}
+			}).AnyTimes()
+
+			for i := 0; i < 5; i++ {
+				res, err := rpcutilClient.OptionTargets([]string{"localhost"}).Ping().Do(ctx)
```

Comment:
Only thing I couldn't figure out, I tried OptionsFactFilter (after populating testdata/facts.yaml)  and it wouldn't discover any results.

Reply:
```
		da, err := discovery.New(srv.AgentManager())
		Expect(err).ToNot(HaveOccurred())
		err=srv.AgentManager().RegisterAgent(ctx, "discover", da, srv.Connector())
		Expect(err).ToNot(HaveOccurred())
```

should sort it, add to `startServerInstance()`

Reply:
The way discovery works is a filtered message is sent to the `discovery` agent with just "ping", and if the server filter validation worked a "pong" is sent back.

https://github.com/choria-io/go-choria/blob/dcc5c3fdb8bbaf2eb736c50c28325e47755afd59/providers/agent/mcorpc/golang/discovery/discovery.go#L57-L72

Agents dont need to be "RPC" agents as such, any payload is supported in any format and any serialization. RPC agents formalize the payload to make a RPC request, but discovery doesnt need all that so just string ping.